### PR TITLE
Add Exchange free/busy provider

### LINKED ISSUE #106: Implement free/busy for Exchange
Based on the great [exchangelib](https://github.com/ecederstrand/exchangelib).

### PR DESCRIPTION
**Pull request by pferreir**
We should see whether `providers/free_busy` is the right place where to have this. I think we could probably group the free/busy "plugins" with the specific authentication ones...

Closes #106

### CODE REVIEW

File: newdle/api.py [hunk 1]
```diff
@@ -119,27 +135,24 @@ def users(q):
...
-        end_time = datetime.time(end).strftime('%H:%M')
-        start_time2 = datetime.time(start2).strftime('%H:%M')
-        end_time2 = datetime.time(end2).strftime('%H:%M')
-        return jsonify([[start_time, end_time], [start_time2, end_time2]])
+def get_busy_times(date, uid):
+    providers = current_app.config.get('FREE_BUSY_PROVIDERS', set())
+    data = []
+
+    for name in providers:
+        module = import_module('newdle.providers.free_busy.{}'.format(name))
```
[the remark this answers is not in the document]
Comment (pferreir) [reply newdle/api.py]: Yeah, modernity!

File: newdle/providers/free_busy/exchange.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    latest_start = max(datetime.combine(day, time(0, 0)), start)
+    earliest_end = min(datetime.combine(day, time(23, 59)), end)
+    diff = (earliest_end - latest_start).days + 1
+    if diff > 0:
+        return (latest_start.time(), earliest_end.time())
+    return None
+
+
+def fetch_free_busy(date, uid):
+    acc = current_app.config.get('EXCHANGE_PROVIDER_ACCOUNT')
```
[the remark this answers is not in the document]
Comment (pferreir) [reply newdle/providers/free_busy/exchange.py]: They are all there, but I commented them out. Does it make sense to provide a default config for an optional module?

File: newdle/client/src/actions.js [hunk 3]
```diff
@@ -173,10 +173,10 @@ export function updateNewdle(newdle) {
   return {type: NEWDLE_UPDATED, newdle};
 }
 
-export function fetchBusyTimesForAnswer(participantEmail, dates) {
+export function fetchBusyTimesForAnswer(participantUid, dates) {
```
[the remark this answers is not in the document]
Comment (pferreir) [reply newdle/client/src/actions.js]: Yeah, me too, just wanted to stay CamelCase-Friendly.

File: newdle/providers/free_busy/exchange.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ThiefMaster) [reply newdle/providers/free_busy/exchange.py]: I'd set them to `None` / empty strings - IMO it's nicer to have exactly one place providing default values for config options. At least Flask extensions usually do that for their options..

[the remark this answers is not in the document]
Comment (pferreir) [reply newdle/providers/free_busy/exchange.py]: OK, that makes sense, thanks!

File: newdle/client/src/answerSelectors.js [hunk 4]
```diff
@@ -119,7 +119,7 @@ export const getNumberOfAvailableAnswers = createSelector(
   getAnswers,
   answers => Object.values(answers).filter(answer => answer === 'available').length
 );
-export const getParticipantEmail = state =>
-  state.answer.participant && state.answer.participant.email;
+export const getParticipantUid = state =>
```
Comment (ThiefMaster) [inline newdle/client/src/answerSelectors.js]: UID

File: newdle/newdle.cfg.example [hunk 5]
```diff
@@ -60,3 +60,21 @@ NEWDLE_CODE_LENGTH = 8
...
+
+# Free/busy provider configuration
+# FREE_BUSY_PROVIDERS = {'exchange'}
+# Random provider: just random data to test the application
+FREE_BUSY_PROVIDERS = {'random'}
+
+# Exchange free/busy configuration
+#
+# This would normally look more or less like:
+# EXCHANGE_DOMAIN = 'acme.com'
```
Comment (ThiefMaster) [inline newdle/newdle.cfg.example]: `example.com`?

File: newdle/providers/free_busy/exchange.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+from marshmallow import fields
+from pytz import utc
+
+
+TYPE_MAP = {'Busy': 'busy', 'Tentative': 'busy', 'OOF': 'busy'}
+
+
+def find_overlap(day, start, end):
+    """Find the overlap of a day with a datetime range."""
+    latest_start = max(datetime.combine(day, time(0, 0)), start)
```
Comment (mic4ael) [inline newdle/providers/free_busy/exchange.py]: `time()` or `time.min`

File: newdle/providers/free_busy/exchange.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    earliest_end = min(datetime.combine(day, time(23, 59)), end)
```
Comment (mic4ael) [inline newdle/providers/free_busy/exchange.py]: maybe `time.end`?

File: newdle/providers/free_busy/exchange.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    configuration = Configuration(
+        server=server, auth_type=NTLM, credentials=credentials
+    )
+
+    uid_account = Account(acc, config=configuration, autodiscover=False)
+    accounts = [
+        (uid_account, 'Organizer', False),
+        (
+            Account(
+                '{}@{}'.format(uid, domain), autodiscover=False, config=configuration
```
Comment (mic4ael) [inline newdle/providers/free_busy/exchange.py]: `f'{uid}@{domain}'`

File: newdle/providers/free_busy/exchange.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+            ),
+            'Optional',
+            False,
+        ),
+    ]
+
+    tz = uid_account.default_timezone
+    start = utc.localize(datetime.combine(date, time(0, 0))).astimezone(tz)
```
Comment (mic4ael) [inline newdle/providers/free_busy/exchange.py]: `time.min`?

Comment (pferreir) [reply newdle/providers/free_busy/exchange.py]: Not sure I undrstand?

**Pull request merged**